### PR TITLE
Enabling a debug dll build under Windows (without CUDA, at least)

### PR DESCRIPTION
**Pull request by MikhailStartsev**
Modified SpaceToDepthOp and DepthToSpaceOp templated classes not to use a SpaceToDepthOpFunctor/DepthToSpaceOpFunctor structs with a template parameter Device=_GPU_Device in case the class itself is instantiated with Device=_CPU_Device. Added a partial template specialization for Device=GPUDevice to preserve the existing behaviour in all cases.

This at least partially (i.e. when bazel is configured not to use CUDA) fixes issue #41118.

### CODE REVIEW

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 1]
```diff
@@ -34,12 +33,14 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/types.h"
 #include "tensorflow/core/util/tensor_format.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
```
Comment (MikhailStartsev) [inline tensorflow/core/kernels/depthtospace_op.cc:36]: The order of includes changed according to clang-formatting tool with `-style=google`

**Comment by MikhailStartsev** [comment]
@rthadur  Re-opening PR #42268 with master branch this time.

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 2]
```diff
@@ -143,6 +127,92 @@ class DepthToSpaceOp : public OpKernel {
   TensorFormat data_format_;
 };
 
+// Template specialization for GPUDevice, explicit referncing GPUDevice in code
```
Comment (sanjoy) [inline tensorflow/core/kernels/depthtospace_op.cc:130]: End comment with period, also typo in `referncing`

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 3]
```diff
@@ -112,23 +113,6 @@ class DepthToSpaceOp : public OpKernel {
...
-      } else if (data_format_ == FORMAT_NCHW) {
-        functor::DepthToSpaceOpFunctor<GPUDevice, T, FORMAT_NCHW> functor;
-        functor(context->eigen_device<GPUDevice>(), Tinput, block_size_,
-                Toutput);
-        return;
-      }
-    }
-
     // NOTE: Assumes data_format_ == FORMAT_NHWC here, since we have rejected
     // (CPU && data_format_ != FORMAT_NHWC) in the constructor.
```
Comment (sanjoy) [inline tensorflow/core/kernels/depthtospace_op.cc:118]: Can we specialize just this part that's different between GPU and CPU?  Same comment for `SpaceToDepthOp`.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 3 above)
Comment (MikhailStartsev) [reply tensorflow/core/kernels/depthtospace_op.cc:118]: That's what I tried to do initially - specialize only the Compute() function, but [AFAIU ](https://stackoverflow.com/a/15374950/1675506) there is no way to specialize part of the class definition only: The template signature of the function has to match that of the class, so to partially specialize the function you need to partially specialize the class. 

Taking the GPU-specialized code into a separate function would result in the same situation - the class still needs to be specialized.

One could probably get around this with some inheritance trickery, but this IMO is making this messier than it should be. Not that I like this code duplication much more...

Comment (sanjoy) [reply tensorflow/core/kernels/depthtospace_op.cc:118]: I mean something like:

```
template <class Device>
struct DepthToSpaceFunctorWrapper {
  void operator(params) {
    // .. generic implementation
  }
};

template <>
struct DepthToSpaceFunctorWrapper<GPUDevice> {
  void operator(params) {
    // .. GPU implementation
  }
};

void Compute(...) {
  // Common stuff ..
  auto Toutput = outputs_tensor->tensor<T, kDims>();
  DepthToSpaceFunctorWrapper<Device>{}(params as needed);
}
```

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 3 above)
Comment (MikhailStartsev) [reply tensorflow/core/kernels/depthtospace_op.cc:118]: Wait, isn't there a much simpler solution? In the if clause [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/depthtospace_op.cc#L115), if one is inside it, it is guaranteed that Device _is_ GPUDevice. So why not replace `functor::DepthToSpaceOpFunctor<GPUDevice, T, FORMAT_NCHW> functor;` by `functor::DepthToSpaceOpFunctor<`**Device**`, T, FORMAT_NCHW> functor;` inside this clause? This should be identical, no?

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 2 above)
Comment (MikhailStartsev) [reply tensorflow/core/kernels/depthtospace_op.cc]: Done

**Comment by MikhailStartsev** [comment]
Please check out the much more concise change version :)

**Review by sanjoy — APPROVED** [review]

**Pull request merged**